File: 2013/S133.cpp
#include <cstdio>
#include <cstring>
int a[2000][10001];
int b[2000];
int main() {
  memset(a, 0, 2000 * 10001 * 4);
  memset(b, 0, 2000 * 4);
  int n,m;
  double ans = 999999.0;
  scanf("%d %d", &n, &m);
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < m; j++) {
      int x, y;
      scanf("%d %d", &x, &y);
      if (x > a[i][y])
        a[i][y] = x;
    }
  }
  for (int i = 1; i <= 10000; i++) {
    double sum = 0.0;
    bool ok = true;
    for (int j = 0; j < n; j++) {
      if (a[j][i] > b[j])
        b[j] = a[j][i];
      if (a[j][i] + b[j] == 0)
        ok = false;
      else
        sum += b[j];
    }
    if (ok && i / sum < ans) 
      ans = i / sum;
  }
  printf("%lf\n", ans);
  return 0;
}

File: 2013/S133.pas
var
  a: array[1..2000, 1..10000] of longint;
  b: array[1..2000] of longint;
  n, m, i, j, x, y: longint;
  ok: boolean;
  sum, ans: double;
begin
  for i := 1 to 2000 do
  begin
    for j := 1 to 10000 do
      a[i][j] := 0;
    b[i] := 0
  end;
  ans := 99999.0;
  read(n, m);
  for i := 1 to n do
    for j := 1 to m do
    begin
      read(x, y);
      if (x > a[i][y]) then
        a[i][y] := x
    end;
  for i := 1 to 10000 do
  begin
    sum := 0.0;
    ok := true;
    for j := 1 to n do
    begin
      if (a[j][i] > b[j]) then
        b[j] := a[j][i];
      if (a[j][i] + b[j] = 0) then
        ok := false
      else
        sum := sum + b[j];
    end;
    if (ok and (i / sum < ans)) then
      ans := i / sum
  end;
  writeln(ans:0:6)
end.
